keps/sig-storage/3141-prevent-volume-mode-conversion/README.md

```diff
@@ -0,0 +1,820 @@
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+A new out-of-tree flag named `PreventVolumeModeConversion` will be introduced on `external-snapshotter`
+and `external-provisioner`. Both of these components are out-of-tree so this proposal will not require any in-tree feature gates. 
+
+### Changes to VolumeSnapshotContent API
+
+With this design, we will introduce two new changes to the VolumeSnapshotContent API:
+1. A new optional field, called `SourceVolumeMode` will be added to the `Spec` of 
+`VolumeSnapshotContents`. This field will be immutable.
+
+```go
+type VolumeSnapshotContentSpec struct {
+...
+// SourceVolumeMode is the mode of the volume whose snapshot is taken.
+// Can be either “Filesystem” or “Block”.
+// If left empty, will be treated as “Unknown”.
+// +optional
+SourceVolumeMode *SourceVolumeMode
+...
+```
+
+2. A new annotation to VolumeSnapshotContent objects. The onus is on the 
+backup vendor (via s/w or manually) to add this annotation to the VolumeSnapshotContent
+if they intend to alter the volume mode during a restore operation.
+
+```go
+snapshot.storage.kubernetes.io/allowVolumeModeChange: true
+```
+
+### Changes to Snapshot Controller
+
+There are two cases to consider:
+1. Dynamic Provisioning
+   1. `VolumeSnapshot` is created by the user, with `VolumeSnapshotClass` optionally
+   specified in the spec.
+   2. `VolumeSnapshotContent` is created by the `snapshot-controller` in response to (i).
+   3. `snapshot-controller` populates the `Spec` of the given `VolumeSnapshotContent`.
+      1. With this change, the controller will fetch the `Spec.PersistentVolumeMode`
+      of the `PV` and add that to newly introduced `Spec.SourceVolumeMode` field of
+      the VolumeSnapshotContent to be created.
+2. Static Provisioning
+   `VolumeSnapshotContent` is created by the admin. With this change, the admin will be 
+expected to fill the `Spec.SourceVolumeMode` field appropriately. If left nil, `Unknown`
+mode will be assumed to preserve existing behavior.
+
+### Changes to external-provisioner
+
+This design leverages the access rights of a user on `VolumeSnapshotContents` to
+determine whether the volume mode can be modified during a restore operation.
+The volume mode can be altered if the requesting user has rights to `Update` and
+`Patch` a`VolumeSnapshotContent` (which is a cluster scoped resource).
+The control flow of a restore operation for a `PVC` from a `VolumeSnapshot` will 
```

Comment:
I think the term "restore operation" may be a little confusing. While creating PVC from VolumeSnapshot can be a restore, it is also typical that during a restore, a new PVC will be created first and then snapshot data will be copied to the new volume.
To avoid confusion, can you just say "The control flow of creating a PVC from a VolumeSnapshot..."

Reply:
Done